Make DWARF debug flag behavior match Clang.

allevato opened the pull request:
Only write the compilation flags to debug info for Mach-O targets, and only
if the RC_DEBUG_OPTIONS environment variable is set.

This change improves hermeticity/reproducibility of Swift binaries/dSYMs by not encoding command line flags that might contain file system paths, since LLDB no longer needs them (except for `-private-discriminator`, which I've left in; it's invariant because it's computed only from the basename anyway).

@adrian-prantl

allevato committed:
Make DWARF debug flag behavior match Clang.

Only write the compilation flags to debug info for Mach-O targets, and only
if the RC_DEBUG_OPTIONS environment variable is set.

adrian-prantl commented on include/swift/Option/FrontendOptions.td:
```diff
@@ -486,4 +486,7 @@ def disable_verify_exclusivity : Flag<["-"], "disable-verify-exclusivity">,
   HelpText<"Enable key path resilience.">;
 
+def use_dwarf_debug_flags : Flag<["-"], "use-dwarf-debug-flags">,
+  HelpText<"Write the compilation flags into the DWARF debug info.">;
```
Perhaps: `-emit-debug-options: Emit the compiler invocation in the debug info.`

allevato replied:
That alternative seems fine; my only concern would be whether it would be easily confused with `-serialize-debugging-options`. Or do you think the symmetry between "serialize" and "emit" would intentionally work well and make it clear which one goes where?

jrose-apple replied:
I don't love either name; it's not the "debug flags" or "debug options" that are being written. (It also doesn't seem DWARF-specific.) I'll toss another contribution in, `-enable-invocation-in-debug-info`, but that's getting fairly wordy.

adrian-prantl replied:
I think it would be nice if the debug-info part came first for grouping and sorting.
`-debug-info-(emit|store)-invocation`, `-debug-info-invocation=yes`?

allevato replied:
If we're making it a driver flag based on the conversation below, should we prefix it with `-g` then (but not add it to the `g_Group`)? How do any of these sound:

* `-gcompiler-flags`
* `-gcompiler-options`
* `-gcompiler-invocation`
* `-grecord-compiler-flags`
* `-grecord-compiler-options`
* `-grecord-compiler-invocation`

jrose-apple replied:
\*sigh\* I'm generally against prefixed options but I think I'm losing that battle. Adrian?

adrian-prantl replied:
Do you have a good reason for why you don't like prefixed options? My main motivation was that they would sort naturally in `swift --help` and `swift (1)` (if we ever got around to actually writing a man page 🙄).

jrose-apple replied:
Sorry, I'm fine with `-debug-info`; it's grouping a bunch of things under single letters that I don't like. To compare with GCC/Clang flags:

- -f* - something to do with the language? Why -f?
- -m* - something to do with code generation? Why -m?
- -d* - something to do with dependencies? Or "defines"?
- -i* - "includes", okay…but it led us to the horrible `-iframework`, meaning "system framework include path"

-g is at least consistently used for debugging things, but there's still no obvious link between "-g" and "debug info"; you just have to memorize it. -O, -W, and -X are pretty much the only flag families that stayed consistent, and -W is still weird around the edges.

jrose-apple replied:
Also we do have a man page! It just tells you to use `-help`. :-)

allevato replied:
Alright, given the feedback so far, if we exclude prefixed flags, I think `-debug-info-store-invocation` is the one that best balances conciseness while still getting across what the flag actually does. `-enable-invocation-in-debug-info` came close, but I agree that it's a bit verbose.

If that sounds good, I'll go ahead and update the PR to use that.

compnerd replied:
@jrose-apple the prefixes actually have some level of consistency

`-f`: feature flags
`-m`: machine flags
`-d`: dump flags (used for the C Preprocessor)
`-I`: includes

jrose-apple replied:
Except that "feature flag" vs. "machine flag" doesn't make any sense to a normal user, and "-i", while consistent, has led to badness elsewhere.

adrian-prantl reviewed (commented):
This is going into the right direction, I added a suggestion for an alternate wording. Let's also hear if @jrose-apple has any opinions.

compnerd commented:
Please don't limit this to Darwin.  This is a spelling issue.  `-grecord-gcc-switches` accomplishes the same thing (it records all of the CC1 flags).

jrose-apple commented:
Definitely in favor of this direction! I've been after Adrian to remove/disable this feature for a while. :-) I'm inclined to make this a driver-level flag too; if it's useful on Darwin it could easily be useful elsewhere, and no one's supposed to use -Xfrontend options in production code (they're not supported or stable). And environment variables are terrible for controlling things like this.

adrian-prantl commented:
> And environment variables are terrible for controlling things like this.
The environment variable is (good design or not) necessary for uniformity with clang on Apple platforms.

I just wanted to point out that the intent of this feature is to help debugging build systems.
(Historically the flags were also used by LLDB during Swift bringup, but that use-case has since been superseded by information serialized in the `.swiftmodule` files.)

compnerd commented:
@allevato - I think as @jrose-apple points out, `-Xfrontend` options aren't as attractive

allevato committed:
Update flag name to -debug-info-store-invocation.

allevato commented:
I've updated the PR to use `-debug-info-store-invocation`, and renamed the toolchain method that lets it override the behavior if the flag isn't present (to preserve the Clang behavior under Darwin, as discussed above). PTAL.

jrose-apple commented on test/DebugInfo/compiler-flags-macosx.swift:
```diff
@@ -0,0 +1,11 @@
+// Check that the sdk and resource dirs end up in the debug info if we build for
+// a Darwin target and set the RC_DEBUG_OPTIONS environment variable. This
+// matches the behavior found in Clang.
+// RUN: %target-swiftc_driver %s -emit-ir -g -target x86_64-apple-macosx10.10 -o - | %FileCheck %s
```
These can just be `%swiftc_driver`, since you're overriding the target anyway.

jrose-apple reviewed (commented):
Looks reasonable to me! Adrian?

jrose-apple commented:
@swift-ci Please test

adrian-prantl commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -1074,8 +1074,10 @@ static void setPrivateDiscriminatorIfNeeded(IRGenOptions &IRGenOpts,
-  if (!PD.empty())
-    IRGenOpts.DWARFDebugFlags += (" -private-discriminator " + PD.str()).str();
+  if (!PD.empty()) {
+    if (!IRGenOpts.DWARFDebugFlags.empty()) IRGenOpts.DWARFDebugFlags += " ";
```
clang-format would format this as:
```
if (!IRGenOpts.DWARFDebugFlags.empty())
   IRGenOpts.DWARFDebugFlags += " ";
```

allevato committed:
Update indentation and %swiftc_driver usage.

adrian-prantl commented:
@swift-ci test

jrose-apple commented:
Oops, of course. By default, we can't compile for macOS on a non-macOS platform. Adding `-parse-stdlib` to that test should get around that by not trying to import Swift.

allevato committed:
Add -parse-stdlib to macOS target test so it builds on Linux.

allevato commented:
Ah-ha, thanks for the tip! Updated.

jrose-apple commented:
@swift-ci Please test

allevato commented:
Hmm, I'm not sure why the test is crashing when emitting the debug info on Linux after the change adding `-parse-stdlib`. It looks similar to this [IRGen/nominal-type-section.swift](https://github.com/apple/swift/blob/master/test/IRGen/nominal-type-section.swift) test, which works; the only meaningful difference I see is that it's invoking the frontend directly instead of the driver. Could that be the important difference here? (I only have a macOS machine to test locally with.)

allevato commented:
Excellent! Thanks for the related fix on the side.
